<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Parallel Taxonomic Classification of Metabarcoding Sequences • BLASTr</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Parallel Taxonomic Classification of Metabarcoding Sequences">
<meta name="description" content="Seamlessly run BLAST searches within R, tailored for ASV (Amplicon Sequence Variant) analysis. This package offers a robust set of functions designed to efficiently identify and process sequences from high-throughput metagenomicmetagenomic experiments, following best practices for sequence data analysis.">
<meta property="og:description" content="Seamlessly run BLAST searches within R, tailored for ASV (Amplicon Sequence Variant) analysis. This package offers a robust set of functions designed to efficiently identify and process sequences from high-throughput metagenomicmetagenomic experiments, following best practices for sequence data analysis.">
<meta property="og:image" content="https://heronoh.github.io/BLASTr/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">BLASTr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/heronoh/BLASTr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="blastr-parallel-taxonomic-classification-of-metabarcoding-sequences-">BLASTr: Parallel Taxonomic Classification of Metabarcoding Sequences 
<a class="anchor" aria-label="anchor" href="#blastr-parallel-taxonomic-classification-of-metabarcoding-sequences-"></a>
</h1>
</div>
<!-- badges: start -->

<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>BLASTr</code> is an R package that seamlessly integrates BLAST+ searches into your R workflow. It is specifically designed for the analysis of Amplicon Sequence Variants (ASVs) from metabarcoding and metagenomic studies. With <code>BLASTr</code>, you can efficiently perform taxonomic classification of your sequences by leveraging the power of parallel processing and automated dependency management.</p>
</div>
<div class="section level2">
<h2 id="features">Features<a class="anchor" aria-label="anchor" href="#features"></a>
</h2>
<ul>
<li>
<strong>Parallel BLAST Searches:</strong> Run multiple BLAST searches concurrently to significantly speed up your analysis.</li>
<li>
<strong>Automated Dependency Management:</strong> <code>BLASTr</code> automatically installs and manages BLAST+ and Entrez Direct dependencies using <code>condathis</code>, ensuring a hassle-free setup.</li>
<li>
<strong>Taxonomic Classification:</strong> Retrieve detailed taxonomic information for your sequences using their NCBI Taxonomy IDs.</li>
<li>
<strong>Flexible and Easy to Use:</strong> The package provides a set of intuitive functions that simplify the process of running thousands of BLAST searches and handling the results.</li>
<li>
<strong>Reproducible Research:</strong> By managing dependencies in isolated Conda environments, <code>BLASTr</code> helps ensure that your analyses are reproducible.</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of <code>BLASTr</code> from GitHub with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"heronoh/BLASTr"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<p>Here’s a simple example of how to use <code>BLASTr</code> to perform a BLAST search and retrieve taxonomic information:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/heronoh/BLASTr" class="external-link">BLASTr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># First, make sure you have the necessary dependencies installed</span></span>
<span><span class="fu"><a href="reference/install_dependencies.html">install_dependencies</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># A vector of ASV sequences</span></span>
<span><span class="va">asvs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"CTAGCCATAAACTTAAATGAAGCTATACTAAACTCGTTCGCCAGAGTACTACAAGCGAAAGCTTAAAACTCATAGGACTTGGCGGTGTTTCAGACCCAC"</span>,</span>
<span>  <span class="st">"CTAGCCATAAACTTAAATGAAGCTATACTAAACTCGTTCGCCAGAGTACTACAAGTGAAAGCTTAAAACTCATAGGACTTGGCGGTGTTTCAGACCCAC"</span>,</span>
<span>  <span class="st">"GCCAAATTTGTGTTTTGTCCTTCGTTTTTAGTTAATTGTTACTGGCAAATGACTAACGACAAATGATAAATTACTAATAC"</span>,</span>
<span>  <span class="st">"AACATTGTATTTTGTCTTTGGGGCCTGGGCAGGTGCAGTAGGAACTTCACTTAGAATAATTATTCGTACTGAGCTTGGGCATCCAGGAAGACTTATCGGGGATGATCAAATCTATAATGTAATTGTTACAGCACATGCATTTGTGATAATTTTTTTTATAGTAATACCTATTATGATT"</span>,</span>
<span>  <span class="st">"ACTATACCTATTATTCGGCGCATGAGCTGGAGTCCTAGGCACAGCTCTAAGCCTCCTTATTCGAGCCGAGCTGGGCCAGCCAGGCAACCTTCTAGGTAACGACCACATCTACAACGTTATCGTCACAGCCCATGCATTTGTAATAATCTTCTTCATAGTAATACCCATCATAATCGGAGGCTTTGGCAACTGACTAGTTCCCCTAATAATCGGTGCCCCCGATATG"</span>,</span>
<span>  <span class="st">"TTAGCCATAAACATAAAAGTTCACATAACAAGAACTTTTGCCCGAGAACTACTAGCAACAGCTTAAAACTCAAAGGACTTGGCGGTGCTTTATATCCAC"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Path to your local FASTA database</span></span>
<span><span class="va">fasta_path</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"BLASTr"</span>, <span class="st">"extdata"</span>, <span class="st">"minimal_db_blast"</span>, ext <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Path to database</span></span>
<span><span class="va">db_path</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/file_temp.html" class="external-link">path_temp</a></span><span class="op">(</span><span class="st">"minimal_db_blast"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/make_blast_db.html">make_blast_db</a></span><span class="op">(</span></span>
<span>  fasta_path <span class="op">=</span> <span class="va">fasta_path</span>,</span>
<span>  db_path <span class="op">=</span> <span class="va">db_path</span>,</span>
<span>  db_type <span class="op">=</span> <span class="st">"nucl"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">fasta_path</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "&gt;AP011979.1 Gymnotus carapo mitochondrial DNA, almost complete genome"</span></span>
<span><span class="co">#&gt; [2] "TACAAACTGGGATTAGATACCCCACTATGCCTAGCCATAAACTTAAATGAAACTATACTAAACTCATTCGCCAGAGTACT"</span></span>
<span><span class="co">#&gt; [3] "ACAAGCGAAAGCTTAAAACTCAAAGGACTTGGCGGTGTTTCAGACCCAC"</span></span>
<span><span class="co">#&gt; [4] "&gt;CP030121.1 Brasilonema octagenarum UFV-E1 chromosome"</span></span>
<span><span class="co">#&gt; [5] "TAGCTCCCGTCGAGTCTCTGCACCTTCCGCATTAGTCATTTATCATTTGTCGTTAGTCATTTGCTAGTAACAATTAACTA"</span></span>
<span><span class="co">#&gt; [6] "AAAACGAAGGACAAAAGACAAATTTGGC"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">db_path</span>, <span class="st">".ndb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run BLAST in parallel</span></span>
<span><span class="va">blast_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/parallel_blast.html">parallel_blast</a></span><span class="op">(</span></span>
<span>  asvs <span class="op">=</span> <span class="va">asvs</span>,</span>
<span>  db_path <span class="op">=</span> <span class="va">db_path</span>,</span>
<span>  total_cores <span class="op">=</span> <span class="fl">2</span> <span class="co"># Number of cores to use</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the taxonomy IDs from the BLAST results</span></span>
<span><span class="va">tax_ids</span> <span class="op">&lt;-</span> <span class="va">blast_results</span><span class="op">$</span><span class="va">`1_staxid`</span></span>
<span></span>
<span><span class="co"># Retrieve taxonomic information in parallel</span></span>
<span><span class="va">taxonomic_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/parallel_get_tax.html">parallel_get_tax</a></span><span class="op">(</span></span>
<span>  organisms_taxIDs <span class="op">=</span> <span class="va">tax_ids</span>,</span>
<span>  total_cores <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  retry_times <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; retrying 0 of 0</span></span>
<span><span class="co">#&gt; ------------------------&gt; unable to retrieve taxonomy for: N/A</span></span>
<span><span class="co">#&gt; ------------------------&gt; unable to retrieve taxonomy for: NA</span></span>
<span><span class="co">#&gt; The following taxIDs could not be retrieved even after 0 attempts:</span></span>
<span><span class="co">#&gt; N/AThe following taxIDs could not be retrieved even after 0 attempts:</span></span>
<span><span class="co">#&gt; NA</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View the results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">blast_results</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 57</span></span>
<span><span class="co">#&gt;   Sequence               `1_subject header` `1_subject` `1_indentity` `1_length`</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                  &lt;chr&gt;              &lt;chr&gt;               &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 CTAGCCATAAACTTAAATGAA… "Gymnotus carapo … AP011979.1           97.0         99</span></span>
<span><span class="co">#&gt; 2 CTAGCCATAAACTTAAATGAA…  &lt;NA&gt;              &lt;NA&gt;                 NA           NA</span></span>
<span><span class="co">#&gt; 3 GCCAAATTTGTGTTTTGTCCT… "Brasilonema octa… CP030121.1           96.2         78</span></span>
<span><span class="co">#&gt; 4 AACATTGTATTTTGTCTTTGG… "Symphoromyia cra… MG967958.1           84.9        179</span></span>
<span><span class="co">#&gt; 5 ACTATACCTATTATTCGGCGC… "Homo sapiens iso… MN849868.1          100          226</span></span>
<span><span class="co">#&gt; 6 TTAGCCATAAACATAAAAGTT… "Hydrochoerus hyd… KX381515.1           99.0         99</span></span>
<span><span class="co">#&gt; # ℹ 52 more variables: `1_mismatches` &lt;dbl&gt;, `1_gaps` &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   `1_query start` &lt;dbl&gt;, `1_query end` &lt;dbl&gt;, `1_subject start` &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   `1_subject end` &lt;dbl&gt;, `1_e-value` &lt;dbl&gt;, `1_bitscore` &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   `1_qcovhsp` &lt;dbl&gt;, `1_staxid` &lt;chr&gt;, `2_subject header` &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   `2_subject` &lt;chr&gt;, `2_indentity` &lt;dbl&gt;, `2_length` &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   `2_mismatches` &lt;dbl&gt;, `2_gaps` &lt;dbl&gt;, `2_query start` &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   `2_query end` &lt;dbl&gt;, `2_subject start` &lt;dbl&gt;, `2_subject end` &lt;dbl&gt;, …</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">taxonomic_info</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 0 × 13</span></span>
<span><span class="co">#&gt; # ℹ 13 variables: Sci_name &lt;chr&gt;, query_taxID &lt;chr&gt;, Superkingdom (NCBI) &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   Kingdom (NCBI) &lt;chr&gt;, Phylum (NCBI) &lt;chr&gt;, Subphylum (NCBI) &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   Class (NCBI) &lt;chr&gt;, Subclass (NCBI) &lt;chr&gt;, Order (NCBI) &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   Suborder (NCBI) &lt;chr&gt;, Family (NCBI) &lt;chr&gt;, Subfamily (NCBI) &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   Genus (NCBI) &lt;chr&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="main-functions">Main Functions<a class="anchor" aria-label="anchor" href="#main-functions"></a>
</h2>
<ul>
<li>
<code><a href="reference/install_dependencies.html">install_dependencies()</a></code>: Installs BLAST+ and Entrez Direct if they are not found on your system.</li>
<li>
<code><a href="reference/make_blast_db.html">make_blast_db()</a></code>: Creates a BLAST database from a FASTA file.</li>
<li>
<code><a href="reference/parallel_blast.html">parallel_blast()</a></code>: Runs BLAST searches for multiple sequences in parallel.</li>
<li>
<code><a href="reference/get_blast_results.html">get_blast_results()</a></code>: Runs a BLAST search for a single sequence.</li>
<li>
<code><a href="reference/parallel_get_tax.html">parallel_get_tax()</a></code>: Retrieves taxonomic information for multiple NCBI Taxonomy IDs in parallel.</li>
<li>
<code><a href="reference/get_tax_by_taxID.html">get_tax_by_taxID()</a></code>: Retrieves taxonomic information for a single NCBI Taxonomy ID.</li>
<li>
<code><a href="reference/run_blast.html">run_blast()</a></code>: A lower-level function to run a BLAST search and return the raw output.</li>
<li>
<code><a href="reference/parse_fasta.html">parse_fasta()</a></code>: Extracts sequences from a FASTA file.</li>
<li>
<code><a href="reference/get_fasta_header.html">get_fasta_header()</a></code>: Retrieves the full header of a sequence from a BLAST database.</li>
</ul>
</div>
<div class="section level2">
<h2 id="dependency-management">Dependency Management<a class="anchor" aria-label="anchor" href="#dependency-management"></a>
</h2>
<p><code>BLASTr</code> uses the <code>condathis</code> package to manage its dependencies (BLAST+ and Entrez Direct). When you run a function that requires one of these tools, <code>BLASTr</code> will automatically check if it’s installed. If not, it will create a Conda environment and install the necessary software. This ensures that you always have the correct versions of the dependencies without having to install them manually.</p>
<p>You can control the installation process with the <code>force</code> and <code>verbose</code> arguments in the <code><a href="reference/install_dependencies.html">install_dependencies()</a></code> and <code><a href="reference/check_cmd.html">check_cmd()</a></code> functions.</p>
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p>Contributions are welcome! Please see the <a href="CONTRIBUTING.html">contributing guide</a> for more details.</p>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>This project is licensed under the MIT License - see the <a href="LICENSE">LICENSE</a> file for details.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/heronoh/BLASTr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/heronoh/BLASTr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing BLASTr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lucio Queiroz <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6090-1834" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Heron Hilario <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2548-1053" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/heronoh/BLASTr/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/heronoh/BLASTr/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lucio Queiroz, Heron Hilario.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
