---
output: gfm
default-image-extension: ""
always_allow_html: true
knitr:
  opts_chunk:
    collapse: true
    comment: "#>"
---

```{r}
#| label: setup-chunk
#| include: false
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# BLASTr: Parallel Taxonomic Classification of Metabarcoding Sequences <a href="https://heronoh.github.io/BLASTr/"><img src="man/figures/logo.png" align="right" height="138" alt="BLASTr website" /></a>

<!-- README.md is generated from README.Rmd. Please edit that file -->

<!-- badges: start -->
[![r-cmd-check](https://github.com/heronoh/BLASTr/actions/workflows/r-cmd-check.yaml/badge.svg)](https://github.com/heronoh/BLASTr/actions/workflows/r-cmd-check.yaml)
<!-- [![CRAN status](https://www.r-pkg.org/badges/version/BLASTr)](https://CRAN.R-project.org/package=BLASTr) -->
<!-- badges: end -->

## Overview

`BLASTr` is an R package that seamlessly integrates BLAST+ searches into your R workflow.
It is specifically designed for the analysis of Amplicon Sequence Variants (ASVs) from metabarcoding and metagenomic studies.
With `BLASTr`, you can efficiently perform taxonomic classification of your sequences by leveraging the power of parallel processing and automated dependency management.

## Features

- **Parallel BLAST Searches:** Run multiple BLAST searches concurrently to significantly speed up your analysis.
- **Automated Dependency Management:** `BLASTr` automatically installs and manages BLAST+ and Entrez Direct dependencies using `condathis`, ensuring a hassle-free setup.
- **Taxonomic Classification:** Retrieve detailed taxonomic information for your sequences using their NCBI Taxonomy IDs.
- **Flexible and Easy to Use:** The package provides a set of intuitive functions that simplify the process of running thousands of BLAST searches and handling the results.
- **Reproducible Research:** By managing dependencies in isolated Conda environments, `BLASTr` helps ensure that your analyses are reproducible.

## Installation

You can install the development version of `BLASTr` from GitHub with:

```{r}
#| label: install-instructions
#| eval: false
# install.packages("devtools")
devtools::install_github("heronoh/BLASTr")
```

## Basic Usage

Here's a simple example of how to use `BLASTr` to perform a BLAST search and retrieve taxonomic information:

```{r}
#| label: setup-blastr
#| eval: false
#| include: false
# Make sure that environments are installed in a temporary directory
withr::local_envvar(
  .new = list(
    `R_USER_DATA_DIR` = fs::path_temp("home", "data")
  )
)

# Create conda environment and install dependencies
BLASTr::install_dependencies(verbose = "silent", force = FALSE)
```

## Obtain NCBI databases

To obtain complet NCBI blast formated databases, procees as follows
```{bash}
#| eval: false

#suggestion: use screen or tmux to emulate a terminal. The downloads usually takes long.
#          tmux: https://tmuxcheatsheet.com/
#          screen: https://kapeli.com/cheat_sheets/screen.docset/Contents/Resources/Documents/index

# download volumes and md5 check files 
seq -w 000 150 | parallel wget https://ftp.ncbi.nlm.nih.gov/blast/db/nt.{}.tar.gz -t 0 --show-progress;
seq -w 000 150 | parallel wget https://ftp.ncbi.nlm.nih.gov/blast/db/nt.{}.tar.gz.md5 -t 0 --show-progress;
wget https://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz -t 0 --show-progress;
wget https://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz.md5 -t 0 --show-progress;
wget https://ftp.ncbi.nlm.nih.gov/blast/db/taxdb-metadata.json -t 0 --show-progress;
     # where 000 is the first volume and 150, the last (up to now).
     
ls *5 | parallel md5sum -c {} >> check.txt
sort check.txt > check_sort.txt

ls *tar.gz | parallel tar -xvzf {} 



  

```

## Making local custom databeses

You can turn any _.fasta_ file with unique headers into a BLAST+ formated database using the _*BLASTr*_ function _make_blast_db_. It is recomended that the unique identifier of each sequence do not contain *.*. You may provide a _.txt_ with the same name as the DB's _.fasta_ file, formated _lcl|_<SequenceId> <TaxonomyId><newline>.

If you want to retrieve _subject Scientific name_  on the results, you can download and extract the *taxdb* files from NCBI, as mentioned above. 

```{r}
#| label: make-blast-db
#| eval: false
library(BLASTr)

# First, make sure you have the necessary dependencies installed
install_dependencies()

# A vector of ASV sequences
asvs <- c(
  "CTAGCCATAAACTTAAATGAAGCTATACTAAACTCGTTCGCCAGAGTACTACAAGCGAAAGCTTAAAACTCATAGGACTTGGCGGTGTTTCAGACCCAC",
  "CTAGCCATAAACTTAAATGAAGCTATACTAAACTCGTTCGCCAGAGTACTACAAGTGAAAGCTTAAAACTCATAGGACTTGGCGGTGTTTCAGACCCAC",
  "GCCAAATTTGTGTTTTGTCCTTCGTTTTTAGTTAATTGTTACTGGCAAATGACTAACGACAAATGATAAATTACTAATAC",
  "AACATTGTATTTTGTCTTTGGGGCCTGGGCAGGTGCAGTAGGAACTTCACTTAGAATAATTATTCGTACTGAGCTTGGGCATCCAGGAAGACTTATCGGGGATGATCAAATCTATAATGTAATTGTTACAGCACATGCATTTGTGATAATTTTTTTTATAGTAATACCTATTATGATT",
  "ACTATACCTATTATTCGGCGCATGAGCTGGAGTCCTAGGCACAGCTCTAAGCCTCCTTATTCGAGCCGAGCTGGGCCAGCCAGGCAACCTTCTAGGTAACGACCACATCTACAACGTTATCGTCACAGCCCATGCATTTGTAATAATCTTCTTCATAGTAATACCCATCATAATCGGAGGCTTTGGCAACTGACTAGTTCCCCTAATAATCGGTGCCCCCGATATG",
  "TTAGCCATAAACATAAAAGTTCACATAACAAGAACTTTTGCCCGAGAACTACTAGCAACAGCTTAAAACTCAAAGGACTTGGCGGTGCTTTATATCCAC"
)

# Path to your local FASTA database
fasta_path <- fs::path_package("BLASTr", "extdata", "minimal_db_blast", ext = "fasta")
# Path to your local Tax ID map file
taxid_map <- fs::path_package("BLASTr", "extdata", "minimal_db_blast", ext = "txt")
# Path to database
db_path <- fs::path_temp("minimal_db_blast")


# BLAST+ requires that the taxid_map file is composede of two columns, separated
#  by a space. The 1st is a unique ID (no . allowed), the first word of the header.
#  It must be preceeded by "lcl|".
#  THe 2nd is the taxID.

make_blast_db(
  fasta_path = fasta_path,
  taxid_map = taxid_map
  db_type = "nucl"
)

head(readLines(fasta_path))

file.exists(paste0(db_path, ".ndb"))

# pra o ssciname funcionar tem que exportar a variavel 
# 
# tem que exportar essa variavel pra funcionar!!!!!
# export BLASTDB=/data/databases/core_nt:/data/databases/core_nt:$BLASTDB



```

```{r}
#| label: blastr-parallel-blast
#| eval: false
# Run BLAST in parallel
blast_results <- parallel_blast(
  query_seqs = asvs,
  db_path = db_path,
  total_cores = 2 # Number of BLAST processes to use
)

blast_results
```

BLASTr keeps track of any errors that may occur during the BLAST searches. You can retrieve the exit codes and STDERR messages using the `exit_codes()` function.

```{r}
#| label: blastr-exit-codes
#| eval: false
# Check for any errors during BLAST searches
exit_code_df <- exit_codes(blast_results)
print(exit_code_df[c("exit_code", "stderr")])
```

```{r}
#| label: blastr-get-taxonomy
#| eval: false
# Extract the taxonomy IDs from the top BLAST hit for each query sequence.
tax_ids <- blast_results$`1_staxid`

# Retrieve taxonomic information in parallel
taxonomic_info <- parallel_get_tax(
  organisms_taxIDs = tax_ids,
  total_cores = 2,
  retry_times = 0
)

print(taxonomic_info)
```

## Main Functions

- `install_dependencies()`: Installs BLAST+ and Entrez Direct if they are not found on your system.
- `make_blast_db()`: Creates a BLAST database from a FASTA file.
- `parallel_blast()`: Runs BLAST searches for multiple sequences in parallel.
- `exit_codes()`: Retrieve the exit codes and STDERR messages from BLAST searches.
- `parallel_get_tax()`: Retrieves taxonomic information for multiple NCBI Taxonomy IDs in parallel.
- `run_blast()`: A lower-level function to run a BLAST search and return the raw output.
- `parse_fasta()`: Extracts sequences from a FASTA file.
- `get_fasta_header()`: Retrieves the full header of a sequence from a BLAST database.

## Dependency Management

`BLASTr` uses the `condathis` package to manage its dependencies (BLAST+ and Entrez Direct).
When you run a function that requires one of these tools, `BLASTr` will automatically check if it's installed.
If not, it will create a Conda environment and install the necessary software.
This ensures that you always have the correct versions of the dependencies without having to install them manually.

You can control the installation process with the `force` and `verbose` arguments in the `install_dependencies()` function.

## Contributing

Contributions are welcome! Please see the [contributing guide](./.github/CONTRIBUTING.md) for more details.

## License

This project is licensed under the MIT License - see the [LICENSE.md](./LICENSE.md) file for details.
